rollup/internal/controller/relayer/l2_relayer_sanity.go

```diff
@@ -0,0 +1,438 @@
+	if prevChunk.EndBlockNumber+1 != chunk.StartBlockNumber {
+		return fmt.Errorf("chunk is not continuous with previous chunk %d: prev end block %d, current start block %d", prevChunk.Index, prevChunk.EndBlockNumber, chunk.StartBlockNumber)
+	}
+
+	// Check L1 messages continuity
+	expectedPoppedBefore := prevChunk.TotalL1MessagesPoppedBefore + prevChunk.TotalL1MessagesPoppedInChunk
+	if chunk.TotalL1MessagesPoppedBefore != expectedPoppedBefore {
+		return fmt.Errorf("L1 messages popped before is incorrect: expected %d, got %d", expectedPoppedBefore, chunk.TotalL1MessagesPoppedBefore)
+	}
+
+	return nil
+}
+
+// validateBlobsAgainstDatabase validates blobs against database records
+func (r *Layer2Relayer) validateBlobsAgainstDatabase(blobs []*kzg4844.Blob, batchesToValidate []*dbBatchWithChunks) error {
+	// Get codec for blob decoding
+	firstBatch := batchesToValidate[0].Batch
+	codec, err := encoding.CodecFromVersion(encoding.CodecVersion(firstBatch.CodecVersion))
+	if err != nil {
+		return fmt.Errorf("failed to get codec: %w", err)
+	}
+
+	// Validate each blob against its corresponding batch
+	for i, blob := range blobs {
+		dbBatch := batchesToValidate[i].Batch
+		dbChunks := batchesToValidate[i].Chunks
+
+		if err := r.validateSingleBlobAgainstBatch(blob, dbBatch, dbChunks, codec); err != nil {
+			return fmt.Errorf("blob validation failed for batch %d: %w", dbBatch.Index, err)
+		}
+	}
+
+	return nil
+}
+
+// validateSingleBlobAgainstBatch validates a single blob against its batch data
+func (r *Layer2Relayer) validateSingleBlobAgainstBatch(blob *kzg4844.Blob, dbBatch *orm.Batch, dbChunks []*orm.Chunk, codec encoding.Codec) error {
+	// Collect all blocks for the batch
+	var batchBlocks []*encoding.Block
+	for _, c := range dbChunks {
+		blocks, err := r.l2BlockOrm.GetL2BlocksInRange(r.ctx, c.StartBlockNumber, c.EndBlockNumber)
+		if err != nil {
+			return fmt.Errorf("failed to get blocks for chunk %d: %w", c.Index, err)
+		}
+
+		if len(blocks) == 0 {
+			return fmt.Errorf("chunk %d has no blocks in range [%d, %d]", c.Index, c.StartBlockNumber, c.EndBlockNumber)
+		}
+
+		// Verify block count matches expected range
+		expectedBlockCount := c.EndBlockNumber - c.StartBlockNumber + 1
+		if len(blocks) != int(expectedBlockCount) {
+			return fmt.Errorf("chunk %d expected %d blocks but got %d", c.Index, expectedBlockCount, len(blocks))
+		}
+
+		batchBlocks = append(batchBlocks, blocks...)
+	}
+
+	// Decode blob payload
+	payload, err := codec.DecodeBlob(blob)
+	if err != nil {
+		return fmt.Errorf("failed to decode blob: %w", err)
+	}
+
+	// Validate L1 message queue hashes
+	if payload.PrevL1MessageQueueHash() != common.HexToHash(dbBatch.PrevL1MessageQueueHash) {
+		return fmt.Errorf("prevL1MessageQueueHash mismatch: decoded=%s, db=%s", payload.PrevL1MessageQueueHash().Hex(), dbBatch.PrevL1MessageQueueHash)
+	}
+
+	if payload.PostL1MessageQueueHash() != common.HexToHash(dbBatch.PostL1MessageQueueHash) {
```

Comment:
Does this catch the case where we wrongly compute the `L1MessageQueueHash` while constructing chunks/batches? I'm afraid that we might just check wrong value against wrong value here

Reply:
Yes. This check cannot defend against wrongly computing `L1MessageQueueHash` error, it just checks that the blob payload's message queue hash values are the same as values stored in DB. 

The check which can detect regression error is in function `validateMessageQueueConsistency`: https://github.com/scroll-tech/scroll/blob/cf852259f36191d55e7d92a77b8fd8a6523d667d/rollup/internal/controller/relayer/l2_relayer_sanity.go#L406-L438.